Register COLLECTIVE_COMM profiler activity type, if available

sanrise opened the pull request:
**Summary:**
Instantiate a collective communications profiler. If a collectives profiler exists then add COLLECTIVE_COMM activity type to the CUDA activity types.

**Test Plan:**
Sample output trace (**internal** use only): https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/clientAPI/0/1707520266/devgpu007.pci1/nccl_activities_2643654_1707520266977.json.gz&bucket=gpu_traces

aaronenyeshi commented on torch/csrc/profiler/kineto_shim.cpp:
```diff
@@ -203,6 +206,24 @@ class ExperimentalConfigWrapper {
+void collectivesProfilerRecord(const std::string& collective_name) {
+#ifdef NCCL_PROXY_PROFILER_ENABLED
+  // if a collective profiler exists and we know that it has the ability to
+  // reord collective names then we can use it here.
```
```suggestion
  // record collective names then we can use it here.
```

aaronenyeshi reviewed (approved):
Looks good, wondering about the two macros: KINETO_HAS_NCCL_PROFILER and NCCL_PROXY_PROFILER_ENABLED. When are either of them set and what differentiates them? Thanks!

aaronenyeshi reviewed (changes requested):
Please check lintrunner failure signal

shengbao-zheng commented:
Just curious if we have to wait this until NCCL_PROXY_PROFILER_ENABLED being upstreamed to NCCL? Otherwise looks like we are using a non-public API where some code is not accessible by all PyTorch users?

sanrise commented:
> Just curious if we have to wait this until NCCL_PROXY_PROFILER_ENABLED being upstreamed to NCCL? Otherwise looks like we are using a non-public API where some code is not accessible by all PyTorch users?

You make a fair point. I should've actually just used the KINETO_HAS_NCCL_PROFILER macro. 

KINETO_HAS_NCCL_PROFILER is only set when Kineto finds any NCCL profiler on startup, so it essentially wraps NCCL_PROXY_PROFILER_ENABLED; so we can just use that and remove this macro usage.

sanrise commented:
@shengbao-zheng I have removed the custom internal API changes to an internal diff at [D54836235] - this one simply registers the correct activity type if Kineto has a NCCL profiler.

sanrise committed:
Register COLLECTIVE_COMM profiler activity type, if available (#121461)

Summary:
**Summary:**
Instantiate a collective communications profiler. If a collectives profiler exists then add COLLECTIVE_COMM activity type to the CUDA activity types.


Test Plan: Sample output trace (**internal** use only): https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/clientAPI/0/1707520266/devgpu007.pci1/nccl_activities_2643654_1707520266977.json.gz&bucket=gpu_traces

Reviewed By: aaronenyeshi, briancoutinho

Differential Revision: D38638806

Pulled By: sanrise

aaronenyeshi reviewed (approved):
Looks great! merging this now

aaronenyeshi commented:
@pytorchbot merge
